ADC/Core/Src/main.c

```diff
@@ -0,0 +1,148 @@
+/*
+ * 	Implement a routine that reads the internal voltage reference (ADC_IN17)
+ * 	and internal temperature sensor (ADC_IN18).
+ * 	The internal reference voltage must be in millivolts and the temperature in Celsius.
+ *
+ * 	1 - Polling		[/]
+ * 	2 - Interrupt	[ ]
+ * 	3 - DMA			[ ]
+ */
+
+#include "stm32l053xx.h"
+#include "macros.h"
+#include "init.h"
+#include "struct.h"
+
+ADC_Status	ADC_Config (uint32_t Channel);	//we call this function with the ADC channel we want to read
+void 		Wait (uint16_t);				//delay function, system clock based
+
+int main (void)
+{
+	uint16_t	measure 	= 0,	//ADC data read is stored here after a conversion
+				v_ref 		= 0;	//internal reference voltage measured (after calculation)
+	int16_t		temperature	= 0;	//temperature in Celsius degrees measured (after calculation)
+
+	ADC_Init();	//ADC initialization function
+
+	while(1)
+	{
+		ADC_Config(ADC_CHSELR_CHSEL18);		//configuring the ADC for internal temperature reading
+		ADC1->CR |= ADC_CR_ADSTART;			//starting the ADC
+		while(!(ADC1->ISR & ADC_ISR_EOC));	//and waiting for End Of Conversion flag to set
+		measure = ADC1->DR;					//storing the data read in measure
+
+		//temperature calculation as in RM example
+		temperature = ((measure * VDD_APPLI / VDD_CALIB) - (int32_t) *TEMP30_CAL_ADDR);
+		temperature = temperature * (int32_t)(130 - 30);
+		temperature = temperature / (int32_t)(*TEMP130_CAL_ADDR - *TEMP30_CAL_ADDR);
+		temperature = temperature + 30;			//the calculated temperature is now stored in temperature
+
+		ADC_Config(ADC_CHSELR_CHSEL17);		//configuring the ADC for internal reference voltage reading
+		ADC1->CR |= ADC_CR_ADSTART;			//starting the ADC
+		while(!(ADC1->ISR & ADC_ISR_EOC));	//and waiting for End Of Conversion flag to set
+		measure = ADC1->DR;					//storing the data read in measure
+
+		//Vref calculation as in RM
+		v_ref = (3 * (int32_t) (*VREFINT_CAL));
+		v_ref = (v_ref * 1000) / measure;		//*1000 so we have Vref in mV
+												//the calculated internal reference voltage is now stored in v_ref
+
+		asm("nop");		//so we can hold the uC here after reading the temperature and Vref
+	}
+
+	return 0;
+}
+
+ADC_Status ADC_Config (uint32_t Channel)
+{
+	if(ADC1->CR & ADC_CR_ADSTART)			//we have to be sure there's no ongoing conversion
+	{
+		ADC1->CR |= ADC_CR_ADSTP;			//if so, we have to set ADSTP bit
+		while(ADC1->CR & ADC_CR_ADSTP);		//and wait for it to clear
+	}
+	ADC1->CR |= ADC_CR_ADDIS;				//disables the ADC
+	while(ADC1->CR & ADC_CR_ADEN);
+	DISABLE_ALL_CHANNELS();					//and disables all channels
```

Comment:
I may be wrong, but do we need to disable the channels? As far as I remember we don't

Reply:
I've got a problem when I tried to read the channels three times or more (reading twice were still ok). I read the temperature first (AN18) and then the internal voltage (AN17), and the ADC is configured to read from AN0 to AN18 (default). What was the problem? Channel selection register with both channels set when I wanted to read only AN18, and since 17 comes first then 18, the ADC read the voltage and applied the measured value to the temperature calc. I don't know why it didn't happen at the second ADC reading, only from the third "in diante"

Reply:
And I was just thinking that I could change that. Reading from channel 18 to 0 or calculating the internal voltage first

Reply:
Probably your problem is with your logical operation. Take a look here: `ADC1->CHSELR |= ADC_CHSELR_CHSEL18;` You are using the operator `|=`instead of `=`. Meaning that the if the register is set to read channel 17, after the operation it's value would be set to read channel 17 and channel 18.

Reply:
--'

Reply:
I've done this mistake several times, know the feel